### PR TITLE
Add API support to configure IdP App Role related configuration

### PR DESCRIPTION
**Pull request by shanggeeth**
## Purpose
- Modify application mgt` PATCH` and `GET` requests to update and view whether an application IdP combination should use application role mappings

- `PATCH` request body -
```
 "appRoleConfigurations": [
        {
            "idp": "Google",
            "useAppRoleMappings": true
        }
    ]
```

- We don't support adding App Role Configurations in application creation payload

## Depends on
- https://github.com/wso2/carbon-identity-framework/pull/4473

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/gen/java/org/wso2/carbon/identity/api/server/application/management/v1/IdpAppRoleConfig.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+        this.idp = idp;
+        return this;
+    }
+    
+    @ApiModelProperty(example = "googleIdP", required = true, value = "")
+    @JsonProperty("idp")
+    @Valid
+    @NotNull(message = "Property idp cannot be null.")
+
```
Comment (ShanChathusanda93) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/gen/java/org/wso2/carbon/identity/api/server/application/management/v1/IdpAppRoleConfig.java:50]:
```suggestion
```

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/gen/java/org/wso2/carbon/identity/api/server/application/management/v1/IdpAppRoleConfig.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    public String getIdp() {
+        return idp;
+    }
+    public void setIdp(String idp) {
+        this.idp = idp;
+    }
```
Comment (ShanChathusanda93) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/gen/java/org/wso2/carbon/identity/api/server/application/management/v1/IdpAppRoleConfig.java:56]: Formatting

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/ServiceProviderToApiModel.java [hunk 3]
```diff
@@ -152,6 +155,21 @@ private AuthenticationSequence buildAuthenticationSequence(ServiceProvider appli
         return authSequence;
     }
 
+    private List<IdpAppRoleConfig> buildIdpAppRoleConfigurations(ServiceProvider application) {
+
+        AppRoleMappingConfig[] applicationRoleMappingConfig = application.getApplicationRoleMappingConfig();
+
+        if (applicationRoleMappingConfig != null) {
```
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/ServiceProviderToApiModel.java:162]: If config is null return empty list

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 4]
```diff
@@ -3212,6 +3216,27 @@ components:
...
+      type: array
+      items:
+        $ref: '#/components/schemas/IdpAppRoleConfig'
+      minItems: 0
+
+    IdpAppRoleConfig:
+      type: object
+      required:
+        - idp
+        - isMappingEnabled
```
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:3229]: useAppRoleMappings

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            for (AuthenticationStep authenticationStep : serviceProvider.getLocalAndOutBoundAuthenticationConfig().
+                    getAuthenticationSteps()) {
+                if (authenticationStep.isAttributeStep()) {
+                        attributeAuthStep = authenticationStep;
+                    break;
+                }
+            }
+        }
+        if (attributeAuthStep != null) {
+            authStepFederatedIdentityProviders = attributeAuthStep.getFederatedIdentityProviders();
```
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java:70]: check if OTP type authenticators related IdPs are also added verify the method

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    private List<String> getAttributeStepFIdPs (ServiceProvider serviceProvider) {
+
+        AuthenticationStep attributeAuthStep = serviceProvider.getLocalAndOutBoundAuthenticationConfig().
+                getAuthenticationStepForAttributes();
+        IdentityProvider[] authStepFederatedIdentityProviders = null;
+        if (attributeAuthStep == null) {
+            for (AuthenticationStep authenticationStep : serviceProvider.getLocalAndOutBoundAuthenticationConfig().
+                    getAuthenticationSteps()) {
+                if (authenticationStep.isAttributeStep()) {
+                        attributeAuthStep = authenticationStep;
```
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java:64]: correct formatting
use stream

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            }
+        }
+        if (attributeAuthStep != null) {
+            authStepFederatedIdentityProviders = attributeAuthStep.getFederatedIdentityProviders();
+        }
+        if (authStepFederatedIdentityProviders != null) {
+            return Arrays.stream(authStepFederatedIdentityProviders).
+                    map(IdentityProvider::getIdentityProviderName).collect(Collectors.toList());
+        }
+        return Lists.newArrayList();
```
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java:76]: use Collections emptylist

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 8]
```diff
@@ -3212,6 +3216,27 @@ components:
...
+      properties:
+        idp:
+          type: string
+          example: googleIdP
+        useAppRoleMappings:
+          type: boolean
+          example: false
+          default: false
+
```
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:3238]: remove empty lines

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    private AppRoleMappingConfig[] getApplicationRoleMappingConfig (List<String> attributeStepFIdPs,
+                                                                    List<IdpAppRoleConfig>  idpAppRoleConfigs) {
```
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java:80]: remove extra space

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+        return attributeStepFIdPs.stream().map(FIdPName -> {
+            AppRoleMappingConfig appRoleMappingConfig = new AppRoleMappingConfig();
```
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java:83]: Have each pipeline in one line

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+     private void updateIdpAppRoleConfigurations(ServiceProvider serviceProvider,
+                                                 List<IdpAppRoleConfig> idpAppRoleConfigurations) {
+
+         List<String> attributeStepFIdPs =  getAttributeStepFIdPs(serviceProvider);
+         AppRoleMappingConfig[] appRoleMappingConfigs = getApplicationRoleMappingConfig(
+                 attributeStepFIdPs, idpAppRoleConfigurations);
+         serviceProvider.setApplicationRoleMappingConfig(appRoleMappingConfigs);
+     }
+
+    private List<String> getAttributeStepFIdPs (ServiceProvider serviceProvider) {
```
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java:55]: remove extra space and proper formatting
check other places

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 12]
```diff
@@ -3212,6 +3216,27 @@ components:
           type: string
           example: basic
 
+    IdpAppRoleConfigurations:
+      type: array
+      items:
+        $ref: '#/components/schemas/IdpAppRoleConfig'
+      minItems: 0
+
+    IdpAppRoleConfig:
```
Comment (shanggeeth) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:3225]: Add comments

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java (hunk 5 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java:79]: OTP type authenticators come under local IdP and will not affect this check.
Attribute step will be set at this point when updating the authentication sequence.

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml (hunk 12 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:3226]: Addressed in bd10a98668ecdf76c803ea61a7ba92d6d28f79cd

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java (hunk 11 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java]: Addressed in bd10a98668ecdf76c803ea61a7ba92d6d28f79cd

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java (hunk 10 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java]: Addressed in bd10a98668ecdf76c803ea61a7ba92d6d28f79cd

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java (hunk 9 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java]: Addressed in bd10a98668ecdf76c803ea61a7ba92d6d28f79cd

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml (hunk 8 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:3242]: Addressed in bd10a98668ecdf76c803ea61a7ba92d6d28f79cd

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml (hunk 4 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml]: Addressed in bd10a98668ecdf76c803ea61a7ba92d6d28f79cd

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java (hunk 7 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java]: Addressed in bd10a98668ecdf76c803ea61a7ba92d6d28f79cd

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/ServiceProviderToApiModel.java (hunk 3 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/ServiceProviderToApiModel.java]: Addressed in bd10a98668ecdf76c803ea61a7ba92d6d28f79cd

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java (hunk 6 above)
Comment (shanggeeth) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java]: Addressed in bd10a98668ecdf76c803ea61a7ba92d6d28f79cd

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        }
+        return Collections.emptyList();
+    }
+
+    /**
+     * Get the application role mapping configurations for the service provider to be updated.
+     *
+     * @param attributeStepFIdPs List of federated identity provider names in the attribute step.
+     * @param idpAppRoleConfigs  List of IdpAppRoleConfig
+     * @return AppRoleMappingConfig[]
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java:93]: complete the comment

Comment (shanggeeth) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/functions/application/UpdateIdpAppRoleConfigurations.java]: Addressed in 7cafe3dda2f22455f4845ec8dae1e99892aa824a

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 14]
```diff
@@ -3315,6 +3319,30 @@ components:
           type: string
           example: basic
 
+    IdpAppRoleConfigurations:
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:3322]: Shall we rename this to `AppRoleConfigurations`?
I feel that `IdpAppRoleConfigurations` is not self explaining

Comment (shanggeeth) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml]: Aligns with the naming in identity framework as well.
Addressed in 8cb8369a51d879d11364fe26a380425f7670bff7

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**